### PR TITLE
Typescript tweaks

### PR DESCRIPTION
**Pull request by lefetmeofefet**
Improved the typescript declarations

### CODE REVIEW

File: src/YoffeeElement.js [hunk 1]
```diff
@@ -49,7 +49,7 @@ class YoffeeElement extends HTMLElement {
         // This is here for when users call `super.connectedCallback()`. Prevent the crash :)
     }
 
-    propUpdatedCallback() {
+    propUpdatedCallback(prop) {
```
Comment (hirako2000) [inline src/YoffeeElement.js:52]: is this change intended?

File: typings/declarations.d.ts [hunk 2]
```diff
@@ -1,36 +1,71 @@
...
      * Returns a `DocumentFragment` that has a one-way binding to each of the objects.
-     * @param {...Object} propsObjects Holds the state of this yoffee element
-     * @return a document fragment
+     * @param {object} propsObjects Objects which are used in the expressions inside the html for updating the html.
+     * @returns {(strings: string[], ...expressionCbs: Function[]) => DocumentFragment} html bound to `propsObjects` as a DocumentFragment.
      */
-    function html(...propsObjects: object[]): any; 
+    function html(...propsObjects: object[]): (strings: TemplateStringsArray, ...expressionCbs: Function[]) => DocumentFragment;
+
+    interface YoffeeElementWrapper extends YoffeeElement {
```
Comment (hirako2000) [inline typings/declarations.d.ts:9]: thanks for complementing the declaration, there were parts I wasn't sure about.

Comment (hirako2000) [reply typings/declarations.d.ts:9]: But those functions may not be exposed publicly, I thought they should not, although of course developers may access them.
I kept the declaration minimal also because I wasn't sure what was intentionally exposed/exported vs what was overlooked.

File: typings/declarations.d.ts [hunk 3]
```diff
@@ -1,36 +1,71 @@
...
-     * @throws an error if the second parameter isn't either a `YoffeeElement` subclass or a `Function`
+     * Creates a new Yoffee element, as a custom web component - a new html tag.
+     * @param elementName The new elements' name. Must include comma.
+     * @param element Declaration of the element: Either a callback function that returns the html, or a `YoffeeElement`
      */
-    function createYoffeeElement(elementName: string, element: Function | YoffeeElement): void;
-    
+    function createYoffeeElement(
+        elementName: string,
+        element: ((props: object, element: YoffeeElementWrapper) => DocumentFragment) | YoffeeElement
```
Comment (hirako2000) [inline typings/declarations.d.ts:34]: I thought it took a function or element, good to know. and this change _makes_ it stronger typed :)

File: src/YoffeeElement.js (hunk 1 above)
Comment (lefetmeofefet) [reply src/YoffeeElement.js:52]: Yes, it received the prop name and now typescript needs to know it.

File: typings/declarations.d.ts [hunk 4]
```diff
@@ -1,36 +1,71 @@
...
+    ): void;
+
     /**
      * @type {YoffeeElement}
      */
-    class YoffeeElement {
+    class YoffeeElement extends HTMLElement {
```
Comment (hirako2000) [inline typings/declarations.d.ts:40]: good catch

File: typings/declarations.d.ts [hunk 5]
```diff
@@ -1,36 +1,71 @@
...
+
+        /**
+         * Called when the element is disconnected from the DOM.
+         */
         disconnectedCallback(): void
-    
+
         /**
-         * / This is here for when users call `super.connectedCallback()`. Prevent the crash :)
+         * Called when the element is connected to the DOM after the element is created
```
Comment (hirako2000) [inline typings/declarations.d.ts:61]: far clearer :)

**Review by hirako2000 — APPROVED** [review]
All good, thanks for the clean up and sync with your more recent changes to yoffee ;)

File: typings/declarations.d.ts (hunk 2 above)
Comment (lefetmeofefet) [reply typings/declarations.d.ts:9]: You're talking about `html` or the `YoffeeElementWrapper`? They're both exposed publicly.
`html` needs to show an error when the expression is not a function so that's crucial.
`YoffeeElementWrapper` helps the user find `onPropUpdate`, `onConnect` and `onDisconnect` callbacks

File: typings/declarations.d.ts (hunk 3 above)
Comment (lefetmeofefet) [reply typings/declarations.d.ts:34]: > I thought it took a function or element

It still does, just defined more strongly as you said :)

File: typings/declarations.d.ts (hunk 2 above)
Comment (hirako2000) [reply typings/declarations.d.ts:9]: talking about `YoffeeElementWrapper.`. But I concede I had not fully understood what it was about :)

Comment (hirako2000) [reply typings/declarations.d.ts:9]: my confusion is down to the fact I *did not* leverage the full capabilities of yoffee yet.

**Comment by lefetmeofefet** [comment]
Only you can merge it, and then I'll merge the pr in the main repo
